Dynamic __all__ defeats mypy strict in downstream packages with no_implicit_reexport

## Problem

esperanto builds its package-level `__all__` at runtime (likely from the factory's plugin registry). This makes the names un-discoverable to static analyzers, so any downstream package using:

```toml
[tool.mypy]
strict = true  # implies no_implicit_reexport = true
```

cannot do:

```python
from esperanto import AIFactory
```

mypy raises something like `error: Module "esperanto" does not explicitly export attribute "AIFactory"`. The workaround is to import from the canonical submodule:

```python
from esperanto.factory import AIFactory
from esperanto.providers.embedding.base import EmbeddingModel
from esperanto.providers.llm.base import LanguageModel
```

This works, but it's not obvious from the README, and a fresh contributor following the docs hits a confusing mypy error before they realize the import path needs to be deeper.

## Reproduction

In a project with `mypy --strict`:

```python
# foo.py
from esperanto import AIFactory  # mypy error: not explicitly exported
```

```bash
$ mypy --strict foo.py
foo.py:1: error: Module "esperanto" does not explicitly export attribute "AIFactory"
```

## Suggestions (any one would help)

1. **Static `__all__`** at the top of `esperanto/__init__.py` listing the public names alongside the dynamic build. Static analyzers see the constant; runtime keeps doing whatever it does.
2. **Ship a `py.typed` marker** + explicit re-exports in `__init__.py` (`from .factory import AIFactory as AIFactory`).
3. **Document the `from esperanto.factory import AIFactory` pattern** prominently in the README so people don't trip on it.

(1) or (2) are the cleanest because they fix the surface; (3) at least documents the gotcha.

## Context

Hit while wiring esperanto into a Python 3.12 monorepo with strict mypy across all packages. Worked around by importing from canonical submodules and it's been stable since. Filing so others don't lose the same hour debugging the import error.

--- a/src/esperanto/__init__.py
+++ b/src/esperanto/__init__.py
@@ -141,37 +141,6 @@ try:
 except ImportError:
     CohereRerankerModel = None  # type: ignore[assignment,misc]
 
-# Store all provider classes
-__provider_classes = {
-    'AnthropicLanguageModel': AnthropicLanguageModel,
-    'GoogleLanguageModel': GoogleLanguageModel,
-    'OpenAILanguageModel': OpenAILanguageModel,
-    'OpenAICompatibleLanguageModel': OpenAICompatibleLanguageModel,
-    'OpenRouterLanguageModel': OpenRouterLanguageModel,
-    'XAILanguageModel': XAILanguageModel,
-    'OpenAIEmbeddingModel': OpenAIEmbeddingModel,
-    'GoogleEmbeddingModel': GoogleEmbeddingModel,
-    "AzureEmbeddingModel": AzureEmbeddingModel,
-    "OllamaEmbeddingModel": OllamaEmbeddingModel,
-    "OllamaLanguageModel": OllamaLanguageModel,
-    "AzureLanguageModel": AzureLanguageModel,
-    "MistralLanguageModel": MistralLanguageModel,
-    "DeepSeekLanguageModel": DeepSeekLanguageModel,
-    "GroqLanguageModel": GroqLanguageModel,
-    "VertexLanguageModel": VertexLanguageModel,
-    "VertexEmbeddingModel": VertexEmbeddingModel,
-    "VertexTextToSpeechModel": VertexTextToSpeechModel,
-    "MistralTextToSpeechModel": MistralTextToSpeechModel,
-    "MistralSpeechToTextModel": MistralSpeechToTextModel,
-    "DeepgramTextToSpeechModel": DeepgramTextToSpeechModel,
-    "CohereLanguageModel": CohereLanguageModel,
-    "CohereEmbeddingModel": CohereEmbeddingModel,
-    "CohereRerankerModel": CohereRerankerModel,
-}
-
-# Get list of available provider classes (excluding None values)
-provider_classes = [name for name, cls in __provider_classes.items() if cls is not None]
-
 # Import factory after defining providers
 from esperanto.factory import AIFactory  # noqa: E402
 
@@ -194,7 +163,29 @@ __all__ = [
     "find_tool_by_name",
     # Profiles
     "OpenAICompatibleProfile",
-] + provider_classes
-
-# Make provider classes available at module level
-globals().update({k: v for k, v in __provider_classes.items() if v is not None})
+    # Provider classes
+    "AnthropicLanguageModel",
+    "GoogleLanguageModel",
+    "OllamaLanguageModel",
+    "OpenAILanguageModel",
+    "OpenAICompatibleLanguageModel",
+    "OpenRouterLanguageModel",
+    "XAILanguageModel",
+    "OpenAIEmbeddingModel",
+    "GoogleEmbeddingModel",
+    "OllamaEmbeddingModel",
+    "AzureEmbeddingModel",
+    "AzureLanguageModel",
+    "MistralLanguageModel",
+    "DeepSeekLanguageModel",
+    "GroqLanguageModel",
+    "VertexLanguageModel",
+    "VertexEmbeddingModel",
+    "VertexTextToSpeechModel",
+    "MistralTextToSpeechModel",
+    "MistralSpeechToTextModel",
+    "DeepgramTextToSpeechModel",
+    "CohereLanguageModel",
+    "CohereEmbeddingModel",
+    "CohereRerankerModel",
+]

--- a/tests/unit/probe_reexport.py
+++ b/tests/unit/probe_reexport.py
@@ -1,0 +1,19 @@
+# ruff: noqa: F401, I001
+from esperanto import (
+    AIFactory,
+    LanguageModel,
+    EmbeddingModel,
+    SpeechToTextModel,
+    TextToSpeechModel,
+    Tool,
+    ToolFunction,
+    ToolCall,
+    FunctionCall,
+    ToolCallValidationError,
+    validate_tool_call,
+    validate_tool_calls,
+    find_tool_by_name,
+    OpenAICompatibleProfile,
+    AnthropicLanguageModel,
+    OpenAILanguageModel,
+)

--- a/tests/unit/test_mypy_reexport.py
+++ b/tests/unit/test_mypy_reexport.py
@@ -1,0 +1,31 @@
+import shutil
+import subprocess
+import sys
+from pathlib import Path
+
+import pytest
+
+
+@pytest.fixture
+def probe_path() -> Path:
+    return Path(__file__).parent / "probe_reexport.py"
+
+
+def test_no_reexport_errors(probe_path: Path) -> None:
+    if shutil.which("mypy") is None:
+        pytest.skip("mypy not on PATH")
+
+    result = subprocess.run(
+        [sys.executable, "-m", "mypy", "--strict", str(probe_path)],
+        capture_output=True,
+        text=True,
+    )
+
+    reexport_errors = [
+        line
+        for line in result.stdout.splitlines()
+        if "does not explicitly export attribute" in line
+    ]
+    assert reexport_errors == [], (
+        "mypy --strict found re-export errors:\n" + "\n".join(reexport_errors)
+    )

